### PR TITLE
Expose necessary interface to guide manual SGD process from Python

### LINKED ISSUE #3959: Python manual sgd
I am trying to implement the SGD functionality to update weights in python manually in caffe python instead of using solver.step() function. The goal is to match the weight updates after doing solver.step() and that by manually updating the weights.

The setup is as follows: Use MNIST data. Set the random seed in solver.prototxt as: random_seed: 52. Make sure momentum: 0.0, weight_decay: 0.0 and, base_lr: 0.01, lr_policy: "fixed". This is done so that, I can simply implement the SGD update equation (with out momentum, regularization etc.). The equation is simply: W_t+1 = W_t - mu \* W_t_diff

Following are the two tests:

**Test1:** Using pycaffe's forward() and backward() to calculate the forward propagation and backward propagation. For each layer that contain weights I do:

```
for k in weight_layer_idx:
    solver.net.layers[k].blobs[0].diff[...] *= lr # weights
    solver.net.layers[k].blobs[1].diff[...] *= lr # biases
```

Next, update the weight/biases as:

```
    solver.net.layers[k].blobs[0].data[...] -= solver.net.layers[k].blobs[0].diff
    solver.net.layers[k].blobs[1].data[...] -= solver.net.layers[k].blobs[1].diff
```

I run this for 5 iterations.

**Test2:** Run caffe's solver.step(5).

Now, what I expect is the two tests should yield exactly same weights after the two iterations.

I save the weights values after each of the above tests and calculate the norm difference between the weight vectors by the two tests, and I see that they are not bit-exact. Can some one spot something that I might be missing?

Following is the entire code for reference:

```
import caffe
caffe.set_device(0)
caffe.set_mode_gpu()
import numpy as np
niter = 5
solver = None
solver = caffe.SGDSolver('solver.prototxt')

# Automatic SGD: TEST2
solver.step(niter)
# save the weights to compare later
w_solver_step = copy(solver.net.layers[1].blobs[0].data.astype('float64'))
b_solver_step = copy(solver.net.layers[1].blobs[1].data.astype('float64'))

# Manual SGD: TEST1
solver = None
solver = caffe.SGDSolver('solver.prototxt')
lr = 0.01

# Get layer types
layer_types = []
for ll in solver.net.layers:
    layer_types.append(ll.type)

# Get the indices of layers that have weights in them
weight_layer_idx = [idx for idx,l in enumerate(layer_types) if 'Convolution' in l or 'InnerProduct' in l]

for it in range(1, niter+1):
    solver.net.forward()  # fprop
    solver.net.backward()  # bprop
    for k in weight_layer_idx:
        solver.net.layers[k].blobs[0].diff[...] *= lr
        solver.net.layers[k].blobs[1].diff[...] *= lr
        solver.net.layers[k].blobs[0].data[...] -= solver.net.layers[k].blobs[0].diff
        solver.net.layers[k].blobs[1].data[...] -= solver.net.layers[k].blobs[1].diff

# save the weights to compare later
w_fwdbwd_update = copy(solver.net.layers[1].blobs[0].data.astype('float64'))
b_fwdbwd_update = copy(solver.net.layers[1].blobs[1].data.astype('float64'))

# Compare
print "after iter", niter, ": weight diff: ", np.linalg.norm(w_solver_step - w_fwdbwd_update), "and bias diff:", np.linalg.norm(b_solver_step - b_fwdbwd_update)
```

The last line that compares the weights with the two tests produces:

after iter 5 : weight diff:  0.000203027766144 and bias diff: 1.78390789051e-05 

where as I expect this difference to be 0.0

Any ideas?

---

Also note, 
If I run these two tests for only 1 iteration, I get exactly matching weight vectors from all layers, but not the subsequent iterations.

### PR DESCRIPTION
**Pull request by mitar**
This fixes #3959 and #2686.

EDIT by @Noiredd: do not merge until #6209 is merged first (or else terrible conflicts will arise during `master` to `opencl` pull).

### CODE REVIEW

File: python/caffe/_caffe.cpp [hunk 1]
```diff
@@ -490,7 +490,8 @@ BOOST_PYTHON_MODULE(_caffe) {
   bp::class_<SolverParameter>("SolverParameter", bp::no_init)
     .add_property("max_iter", &SolverParameter::max_iter)
     .add_property("display", &SolverParameter::display)
-    .add_property("layer_wise_reduce", &SolverParameter::layer_wise_reduce);
+    .add_property("layer_wise_reduce", &SolverParameter::layer_wise_reduce)
+    .add_property("base_lr", &SolverParameter::base_lr, &SolverParameter::set_base_lr);
```
Comment (Noiredd) [inline python/caffe/_caffe.cpp]: This change is included in #6209 which will be merged prior to any other changes to `_caffe.cpp`

File: python/caffe/_caffe.cpp [hunk 2]
```diff
@@ -507,13 +508,15 @@ BOOST_PYTHON_MODULE(_caffe) {
...
     .add_property("param", bp::make_function(&Solver<Dtype>::param,
-              bp::return_value_policy<bp::copy_const_reference>()));
+              bp::return_internal_reference<>()));
   BP_REGISTER_SHARED_PTR_TO_PYTHON(Solver<Dtype>);
 
   bp::class_<SGDSolver<Dtype>, bp::bases<Solver<Dtype> >,
     shared_ptr<SGDSolver<Dtype> >, boost::noncopyable>(
-        "SGDSolver", bp::init<string>());
+        "SGDSolver", bp::init<string>())
+        .def("get_learning_rate", &SGDSolver<Dtype>::GetLearningRate);
```
Comment (Noiredd) [inline python/caffe/_caffe.cpp]: What about other solvers, will they get this property only due to the inheritance?

File: src/caffe/solver.cpp [hunk 3]
```diff
@@ -266,10 +266,6 @@ void Solver<Dtype>::Step(int iters) {
     }
     ApplyUpdate();
 
-    // Increment the internal iter_ counter -- its value should always indicate
-    // the number of times the weights have been updated.
-    ++iter_;
-
```
Comment (Noiredd) [inline src/caffe/solver.cpp]: Why did this have to be moved to `SGDSolver`?

File: python/caffe/_caffe.cpp (hunk 1 above)
Comment (mitar) [reply python/caffe/_caffe.cpp]: Oho, looking at #6209 looks like everything in this pull request is done also there, no?

Comment (Noiredd) [inline python/caffe/_caffe.cpp]: This change is included in #6209 which will be merged prior to any other changes to `_caffe.cpp`.

**Comment by mitar** [comment]
Thanks for the quick code review!

File: python/caffe/_caffe.cpp (hunk 2 above)
Comment (Noiredd) [inline python/caffe/_caffe.cpp]: What about other solvers, will they get this property only due to the inheritance?

File: src/caffe/solver.cpp (hunk 3 above)
Comment (Noiredd) [inline src/caffe/solver.cpp]: Why did this have to be moved to SGDSolver?

**Comment by Noiredd** [comment]
Please run `make lint` and fix the style issues and we will resume the review.  
Also, I'm not sure if you noticed but I accidentally removed your comment - please repost it if needed.

File: python/caffe/_caffe.cpp (hunk 1 above)
Comment (mitar) [reply python/caffe/_caffe.cpp]: Oho, looking at #6209 looks like everything in this pull request is done also there, no?

File: python/caffe/_caffe.cpp (hunk 2 above)
Comment (mitar) [reply python/caffe/_caffe.cpp]: Other solvers extending from `SGDSolver`? I think only them will get this, but this is OK, no? this is on purpose.

File: src/caffe/solver.cpp (hunk 3 above)
Comment (mitar) [reply src/caffe/solver.cpp]: Based on the comment: "its value should always indicate the number of times the weights have been updated."

Because now it is possible to update weights without having to call `step`, this has to be moved so that also calling directly `ApplyUpdate` increases it. I think that this is even cleaner from the design perspective. The code is next to the code which is updating weights. Not somewhere else.

**Comment by mitar** [comment]
> Good addition, I was about to implement something like this myself but first we need to pull some Pycaffe changes from the OpenCL branch to synchronize the whole repository. For this reason I recommend holding this until #6209 is reviewed and merged.

Thanks. I can rebase after that is merged.

> Please run make lint and fix the style issues and we will resume the review.

I cannot run it, it says:

```
Makefile:6: *** Makefile.config not found. See Makefile.config.example..  Stop.
```

Also, I cannot see results from Travis CI run. Are they set as private?

File: python/caffe/_caffe.cpp (hunk 2 above)
Comment (Noiredd) [reply python/caffe/_caffe.cpp]: So if I do `a = caffe.AdamSolver("solver.prototxt")` in my pycaffe script, will I be able to call `a.get_learning_rate()` or not?

**Comment by mitar** [comment]
Pushed linting fixes.

File: python/caffe/_caffe.cpp (hunk 2 above)
Comment (mitar) [reply python/caffe/_caffe.cpp]: To my understanding, the answer is yes, but I have not tested it.

Comment (Noiredd) [reply python/caffe/_caffe.cpp]: Would be good to know this. Maybe consider implementing some [tests](https://github.com/BVLC/caffe/blob/master/python/caffe/test/test_solver.py) for this change? See #4342, it is somewhat similar to this (a bit more clumsy) but also had some very simple tests.

**Comment by shelhamer** [comment]
Thanks @mitar! I have been meaning to make a PR like this myself for some time. I'll take a look at both #6209 and this soon.

File: src/caffe/solver.cpp (hunk 3 above)
Comment (shelhamer) [reply src/caffe/solver.cpp]: Ideally the base `Solver` would enforce the increment to `iter_`, but since all Caffe solvers are SGD solvers and other solvers that depart from this mold aren't expected I think the move is a pragmatic choice that avoids unnecessary interface code.

File: python/caffe/_caffe.cpp (hunk 2 above)
Comment (shelhamer) [reply python/caffe/_caffe.cpp]: No, this does not inherit correctly b.c. the `bp::bases` of the child `SGDSolver`s are set to `Solver` and not `SGDSolver`. Switching the `NesteroveSolver` and other children to have the `SGDSolver` base fixes the issue. I second incorporating the `apply_update()` test from c5b5b55845c20930026f1f563136247c41f58075, so could you cherry-pick it to give @nitnelave credit for it?

**Comment by shelhamer** [comment]
@mitar I had time to do a pass so I updated this branch by rebasing on master, rewording for detail, and including a test for updating from pycaffe. I'm planning to merge once the tests pass to double-check my local results.

**Comment by mitar** [comment]
Oh, thanks!

**Comment by mitar** [comment]
What are plans for a release which would include this? Or do you have some daily build I could pip install without compiling the repo myself?

**Comment by shelhamer** [comment]
The current state of affairs with master is that you have to compile it yourself.

> What are plans for a release which would include this?

There is no definite plan right now but I could imagine a release in early July to collect improvements since 1.0. I will bring it up with the core devs.

> Or do you have some daily build I could pip install without compiling the repo myself?

No, we do not have a daily build, although it would obviously be useful. I've never rigged up such a system myself, but perhaps a daily linux build (specifically for the Ubuntu LTS) would be feasible.

**Comment by mitar** [comment]
> No, we do not have a daily build, although it would obviously be useful. I've never rigged up such a system myself, but perhaps a daily linux build (specifically for the Ubuntu LTS) would be feasible.

You could even push artifacts from Travis CI.

I think you also publish a PyPi package like other libraries do. Using [manylinux](https://github.com/pypa/manylinux) and if you need more space there for the package, [you can request it here](https://github.com/pypa/packaging-problems).
